### PR TITLE
Add new private `vips` package

### PR DESCRIPTION
**Pull request by swissspidy**
## What?

This is a new package for client-side media processing, see #61447.

It is a wrapper around [wasm-vips](https://github.com/kleisauke/wasm-vips), which is a WASM port of the powerful vips image library.

This package has been tested very well as part of https://github.com/swissspidy/media-experiments, where it runs in a dedicated web worker.

The package is private and shall not be published until later on when all of the subsequent PRs are merged.

## Why?

Required for browser-based image editing.

## How?

Adds a new lightweight package for using vips in a browser.

It's a separate package so it can be easily loaded in a (inline) web worker. That's also why there is this `setLocation` method, to ensure that works in a WordPress context.

## Testing Instructions

None at this time, as this is not user-facing or anything.

There are some tests which should pass of course.

### Testing Instructions for Keyboard

N/A

## Screenshots or screencast 

N/A

### CODE REVIEW

File: tools/webpack/shared.js
Comment (swissspidy) [inline tools/webpack/shared.js]: Looks like this has implications for other packages as well :/

**Comment by swissspidy** [comment]
@youknowriad @ntsekouras Would appreciate your early thoughts on this one

The `vips` package is used by the upcoming `upload-media` package (all coming from https://github.com/swissspidy/media-experiments), which loads it in an _inline_ web worker.

It is a wrapper around `wasm-vips` which — surprise — contains `.wasm` files. I used to load these from a CDN before, but they should really be served from the local server instead. I tried to update the webpack config to make that work.

There are currently two big issues with that:

1. I need to ensure the `wasm` files are part of the build output everywhere, as the JS files need to load them.
2. I need to load the CJS version of `wasm-vips`, not the ESM version.
Fiddling with [`conditionNames`](https://webpack.js.org/configuration/resolve/#resolveconditionnames) works but seems to have unintended effects on other packages' imports

The whole thing is best tested once I add the PR for the `upload-media` package, but for that I need some of my other PRs to be merged first :-)

File: packages/vips/package.json
Comment (sirreal) [inline packages/vips/package.json]: This seems new, I don't think any other packages use it. That's not a bad thing and we're considering using exports for building more script modules. CC @gziolo 👀 

Since this is unused so far in WordPress packages, I'd prefer not to introduce it here. It can be added to all packages at the same time in the same way when we decide to do that.

---

I am surprised to see it point to a TypeScript file. I think this should point to a compiled file, probably the same as the `module` field:

```suggestion
	"exports": {
		".": {
			"default": "./build-module/index.js"
		}
	},
```

When the field is defined like this with only a single `.` export, there's a shorthand form: `"exports": "./build-module/index.js"`.

Comment (swissspidy) [reply packages/vips/package.json]: This is mostly leftover from testing I think, agree it should be removed for consistency :-)

But my main focus right now is getting the WASM thing to work 😄

**Comment by sirreal** [comment]
Will you provide some details about the build issues? In particular how should the behavior be verified. Is a successful `npm run build` sufficient?

- I understand that the changes in `tools/webpack/packages.js` should be updated so they're not package-specific. Is there anything else with that file?
- Will you explain this comment in the shared webpack config https://github.com/WordPress/gutenberg/pull/64845/files?show-viewed-files=true&file-filters%5B%5D=#r1734286736? I'd like to understand why that was added.

**Comment by swissspidy** [comment]
> Will you provide some details about the build issues? In particular how should the behavior be verified. Is a successful `npm run build` sufficient?

As per the PR description, after a build, the `.wasm` files should appear in `build/vips`. Ideally without hardcoding.
I think they need to be in `packages/vips/build` and `packages/vips/build-module` as well, because e.g. WP core will need to be able to load those files as well.

I'll see if I can add some simple demo or so to verify the behavior more easily in the browser / editor 👍 

> * I understand that the changes in `tools/webpack/packages.js` should be updated so they're not package-specific. Is there anything else with that file?

I don't know. Is this file also used for creating the `packages/*/build` folders? As per the above, I think I need to include the `.wasm` files in there as well.

> * Will you explain this comment in the shared webpack config [#64845 (files)](https://github.com/WordPress/gutenberg/pull/64845/files?show-viewed-files=true&file-filters%5B%5D=#r1734286736)? I'd like to understand why that was added.

As per https://github.com/WordPress/gutenberg/pull/64845#issuecomment-2314760962, I need to load the CJS version of `wasm-vips`, not the ESM version. Fiddling with [conditionNames](https://webpack.js.org/configuration/resolve/#resolveconditionnames) works but seems to have unintended effects on other packages' imports.

See the `exports` of `wasm-vips`:

https://github.com/kleisauke/wasm-vips/blob/87958eda7f54cdc04fbe9991a3068dbb74b05737/package.json#L16-L33

With the change to `conditionNames` it will load `vips.js` instead of the `vips-es6.js` file

**Comment by sirreal** [comment]
Will you share some details on the unintended consequences for other packages? Is it breaking some behavior or how does it manifest?

**Comment by swissspidy** [comment]
> Will you share some details on the unintended consequences for other packages? Is it breaking some behavior or how does it manifest?

The bundle size increases for `build/components/index.min.js` and `build/editor/index.min.js` are very suspicious. Because the `conditionNames` change affects _all_ packages loaded by webpack, I assume `editor` and `components` now load different (bigger) files for some of their dependencies.

**Comment by swissspidy** [comment]
@sirreal I just pushed a bunch of changes to make it easier to test the PR. I will remove those changes again before merge.

You can test this by dropping an image file into the editor. It should upload normally, but behind the scenes it will be compressed by vips.

The setup is similar to the desired end state with the new client-side media processing. `@wordpress/vips` will be dynamically loaded in an inline worker using https://www.npmjs.com/package/@shopify/web-worker. That's why it requires `__webpack_public_path__`, so the worker knows where to find the WASM files.

Note: We probably don't want to use `__webpack_public_path__` directly in published packages, so we could use something like `globalThis.PUBLIC_PATH` & `DefinePlugin` or so in the final implementation.

**Comment by sirreal** [comment]
This is a mess 😖 

In theory, webpack supports wasm natively by enabling `experiments.asyncWebAssembly`. I also switched from using `conditionNames` (affecting all packages) to patch the `wasm-vips` package. This still didn't get things working.

<details>

<summary>patch with some ideas</summary>

```diff
diff --git before/patches/wasm-vips+0.0.10.patch after/patches/wasm-vips+0.0.10.patch
new file mode 100644
index 00000000000..3310b706e5a
--- /dev/null
+++ after/patches/wasm-vips+0.0.10.patch
@@ -0,0 +1,15 @@
+diff --git a/node_modules/wasm-vips/package.json b/node_modules/wasm-vips/package.json
+index c20f052..b26a38a 100644
+--- a/node_modules/wasm-vips/package.json
++++ b/node_modules/wasm-vips/package.json
+@@ -15,10 +15,6 @@
+   "type": "commonjs",
+   "exports": {
+     ".": {
+-      "browser": {
+-        "import": "./lib/vips-es6.js",
+-        "require": "./lib/vips.js"
+-      },
+       "node": {
+         "import": "./lib/vips-node.mjs",
+         "require": "./lib/vips-node.js"
diff --git before/tools/webpack/packages.js after/tools/webpack/packages.js
index d4bf1243c8f..6fd792e424f 100644
--- before/tools/webpack/packages.js
+++ after/tools/webpack/packages.js
@@ -143,20 +143,6 @@ module.exports = {
 		// This is the default, but required for @shopify/web-worker.
 		globalObject: 'self',
 	},
-	module: {
-		rules: [
-			...baseConfig.module.rules,
-			{
-				test: /\.wasm$/,
-				type: 'asset/resource',
-				generator: {
-					// FIXME: Do not hardcode path.
-					filename: './build/vips/[name].wasm',
-					publicPath: '',
-				},
-			},
-		],
-	},
 	performance: {
 		hints: false, // disable warnings about package sizes
 	},
diff --git before/tools/webpack/shared.js after/tools/webpack/shared.js
index bad55ca53cd..51aba37dee0 100644
--- before/tools/webpack/shared.js
+++ after/tools/webpack/shared.js
@@ -16,6 +16,9 @@ const { NODE_ENV: mode = 'development', WP_DEVTOOL: devtool = 'source-map' } =
 
 const baseConfig = {
 	target: 'browserslist',
+	experiments: {
+		asyncWebAssembly: true,
+	},
 	optimization: {
 		// Only concatenate modules in production, when not analyzing bundles.
 		concatenateModules:
@@ -48,13 +51,6 @@ const baseConfig = {
 			},
 		],
 	},
-	resolve: {
-		// Ensure "require" has a higher priority when matching export conditions.
-		// Needed for wasm-vips which is used by the @wordpress/vips package.
-		// https://webpack.js.org/configuration/resolve/#resolveconditionnames
-		// https://github.com/kleisauke/wasm-vips/issues/50#issuecomment-1664118137
-		conditionNames: [ 'require', 'import' ],
-	},
 	watchOptions: {
 		ignored: [
 			'**/node_modules',
```

</details>

**Could this be used as a Script Module?** I haven't tried yet, but if it were compiled as an actual module it might solve some of these issues. I'm working on building some more script modules in #65064.

**Comment by swissspidy** [comment]
> This is a mess 😖

Sort of... At least right now, the PR in its current form, the WASM files were correctly loaded, so it did seem to work as expected.

> I also switched from using `conditionNames` (affecting all packages) to patch the `wasm-vips` package. This still didn't get things working.

What exactly isn't working? Does that break the build step?

(Aside: the wasm-vips maintainer is super responsive, in case we need to make any upstream changes)

> In theory, webpack supports wasm natively by enabling `experiments.asyncWebAssembly`.
> **Could this be used as a Script Module?** I haven't tried yet, but if it were compiled as an actual module it might solve some of these issues. I'm working on building some more script modules in #65064.

Are script modules really the answer? 😄  That opens a whole other can of worms IMHO. This library needs to be called from the editor, but that isn't a script module.

**Comment by sirreal** [comment]
I've spent some time fighting with this again and I've made progress, but I'm still not happy with it. I'm starting to think the necessary wasm files should just be copied with the `CopyWebpackPlugin`.

---

When I said this was a mess, I was referring to the build situation and nothing particular in this PR. I hope that was clear 🙇 

---

> Are script modules really the answer?

Maybe, they might improve some of the problems we're experiencing in this PR. I'd like to experiment a bit and see.

We can find ways to allow scripts to depend on script modules. It's trivial to _use_ modules from scripts via `import()` (although it requires promises/async code) and technically it should be straightforward. I'd really like to move that work ahead: https://core.trac.wordpress.org/ticket/61500

In general, I'd like it if new packages were introduced as script modules and not scripts.

---

I understand the problem as something like this:

Webpack is a (module) bundler. It's happy to transform modules into other forms, and modules are the reason that it uses the `vips-es6.js` file from the package. The problem is that it compiles that file into a script, where `import.meta.url` isn't valid syntax, which breaks when the generated script is later parsed.

Most other places it's fine to use modules because they're mostly just using import/export syntax and not other parts of the module specification.

**Comment by sirreal** [comment]
One of the ways this breaks when loading the wasm files as wasm is errors like this:

```
ERROR in ./node_modules/wasm-vips/lib/vips-jxl.wasm
Module not found: Error: Can't resolve 'GOT.func' in '…/node_modules/wasm-vips/lib'
```

Maybe loading them as assets is fine 🤷

File: tools/webpack/shared.js
Comment (sirreal) [reply tools/webpack/shared.js]: There's a package patch in https://github.com/WordPress/gutenberg/pull/65250/files?show-viewed-files=true&file-filters%5B%5D=#diff-57cf15f5ef3f76d6b8f1dcfe9ca40e86887fd295f80df3fb2bc0565352520d31 that can be used instead of this change.

File: tools/webpack/packages.js
Comment (sirreal) [inline tools/webpack/packages.js:153]: This may be alright, maybe it should go under `build/vendor/[name].wasm` and it should match wasm in the `wasm-vips` package.

Comment (swissspidy) [reply tools/webpack/packages.js:153]: Moving to `build/vendors` feels wrong though 🤔 

Especially since it should also be in `packages/vips/build` and `packages/vips/build-module`

File: babel.config.js
Comment (swissspidy) [inline babel.config.js]: Note to self: need to remove this plugin for Jest tests

**Comment by swissspidy** [comment]
> I've spent some time fighting with this again and I've made progress, but I'm still not happy with it. I'm starting to think the necessary wasm files should just be copied with the `CopyWebpackPlugin`.

I can definitely give that a try to see if it works.

RIght now the package is private anyway, so we can easily iterate on this.

Proposed next steps

1. Clean up this PR, ensure that tests pass and everything works as expected
2. Extract unrelated changes (e.g. the PHP changes, the web worker implementation I added) into new PRs
3. Merge this PR
4. Profit!

**Comment by sirreal** [comment]
I plan to help sort out the details here, but I'm prioritizing other work I hope to ship in WordPress 6.7. After that work is settled I can dedicate some more focus to this. Some of that work is related, such as how we're building and shipping Script Modules: https://github.com/WordPress/gutenberg/pull/65460

> Extract unrelated changes (e.g. the PHP changes, the web worker implementation I added) into new PRs

This is always good if we can land the parts that are clear and postpone the parts that are complicated 👍

**Comment by swissspidy** [comment]
New PR for the PHP changes: #65701

**Comment by swissspidy** [comment]
I removed all the web worker stuff now from this PR to keep this to the bare minimum, unblocking further work on this feature.

@sirreal I'd appreciate another review here, just focusing on the existing code. The real fun begins (again) in subsequent PRs when actually using the wasm files in a worker.

File: packages/vips/src/types.ts
Comment (sirreal) [inline packages/vips/src/types.ts]: What's `@mextp/upload-media`? It doesn't seem to be a public npm package.

File: packages/vips/src/types.ts
Comment (sirreal) [inline packages/vips/src/types.ts]: It jumped out at me that one place uses `center` and another `centre`. I think `centre` comes from lib vips interface. Could the package use one spelling and adjust internally?

**Review by sirreal — COMMENTED** [review]
I did a quick scan and have some minor notes to share. I'll come back and do a proper review sometime this week.

File: packages/vips/src/types.ts
Comment (swissspidy) [reply packages/vips/src/types.ts]: `mexp` is https://github.com/swissspidy/media-experiments, where this code is from :-)

I'll need to remove this comment

File: packages/vips/src/types.ts
Comment (swissspidy) [reply packages/vips/src/types.ts]: Yeah vips uses different spelling. But the `Interesting` type with `centre` here is not exported, it's all used internally. Used here:

https://github.com/WordPress/gutenberg/blob/1e90ef282524a82c5daa537fe1315415d243bc51/packages/vips/src/index.ts#L264

File: test/unit/config/global-mocks.js
Comment (sirreal) [inline test/unit/config/global-mocks.js:51]: Can these be set conditionally like TextDecoder and TextEncoder above?

```suggestion
if ( ! global.Blob ) {
	global.Blob = BlobPolyfill;
}
if ( ! global.File ) {
	global.File = FilePolyfill;
}
```

File: test/unit/config/wasm-stub.js
Comment (sirreal) [inline test/unit/config/wasm-stub.js]: Why does this have a process function? I'm not well versed in WASM and it's not clear to me if this is simulating wasm module behavior or if this is something specific to the vips package.

File: packages/vips/package.json
Comment (sirreal) [inline packages/vips/package.json:33]: Above this is `private: true` - I think that will not allow it to be published. I understand this is _not_ intended to be published to npm at this time.

File: packages/vips/src/index.ts
Comment (sirreal) [inline packages/vips/src/index.ts]: There are a number of places throughout this PR that use type alias `type T = {}` instead of interfaces `interface I {}`.

Can we switch to interface? [TypeScript has a history of recommending this:](https://www.typescriptlang.org/docs/handbook/advanced-types.html#interfaces-vs-type-aliases)

> Because an interface more closely maps how JavaScript objects work [by being open to extension](https://wikipedia.org/wiki/Open/closed_principle), we recommend using an interface over a type alias when possible.

File: packages/vips/src/types.ts
Comment (sirreal) [inline packages/vips/src/types.ts:8]: It took me a while to see, but this is a 2-tuple like `[ 'left', 'bottom' ]` or `[ 'center', 'center' ]`.

File: packages/vips/src/types.ts
Comment (sirreal) [inline packages/vips/src/types.ts]: "force" is undoumented here.

**Review by sirreal — COMMENTED** [review]
I've been reviewing this in depth, I'm leaving another round of feedback. I haven't finished reviewing the main TS files yet.

I'd strongly prefer that if and when this arrives in WordPress it's as a _script module_. I know there are pieces missing for that but they're things I'm working on and I think we can have for 6.8.

On Monday I'll finish this complete review. Thanks for your patience 🙂

File: packages/vips/src/types.ts
Comment (swissspidy) [reply packages/vips/src/types.ts]: Good catch. Relevant type: https://github.com/kleisauke/wasm-vips/blob/7b3657e0e2b45728b4947c577f0b75147570b627/lib/vips.d.ts#L2703-L2726

File: packages/vips/package.json
Comment (swissspidy) [reply packages/vips/package.json:33]: Correct.

**Review by sirreal — APPROVED** [review]
I've reviewed this and it seems ready. There are some remaining questions and minor suggestions from my previous review https://github.com/WordPress/gutenberg/pull/64845#pullrequestreview-2378173811, but nothing blocking.

File: test/unit/config/global-mocks.js
Comment (swissspidy) [reply test/unit/config/global-mocks.js:51]: I don't think so, because this overrides the default `Blob` implementation provided by jsdom, which is incomplete. Node on the other hand has the full thing already baked in.

Conditionally adding them causes tests to fail again.

File: test/unit/config/wasm-stub.js
Comment (swissspidy) [reply test/unit/config/wasm-stub.js]: I think the contents don't actually matter, from what I can see it could just be an empty file. So I guess I'll make it empty with a comment or so.

File: tools/webpack/packages.js
Comment (youknowriad) [inline tools/webpack/packages.js:157]: Can someone explain to me how we're building this. What's the story here? We just copy the "wasm" files into the built folders? What does webpack outputs when it finds a wasm dependency? What about the npm packages that we publish, are we changing the constraints for developers using these npm packages if these packages start to require wasm dependencies? 

Basically, I'm trying to understand the impact on:

 - npm consumers.
 - WordPress scripts.

Comment (sirreal) [reply tools/webpack/packages.js:157]: It's configured as a webpack "asset", similar to images and other assets. Webpack basically copies the file and exposes the path when the asset is imported.

It shouldn't be published to npm for now, so I don't think that's a concern at this time.

For WordPress scripts it should work because we can figure out the path that needs to be used to access the wasm assets.

These wasm assets are one reason I'd really like to use script modules for this package. Webpack has some wasm module support and the library relies on some modules features, so there may be some simplification available.

Comment (youknowriad) [reply tools/webpack/packages.js:157]: > It shouldn't be published to npm for now, so I don't think that's a concern at this time.

For me that's a concern, my guess is that we're going to add the `vips` packages as a dependency of `block-editor` package. So my question is what happens to third-party block editors, will they have to change how their applications builds, feels heavy? How do regular npm packages (in the npm registry) that use wasm dependencies work.

Comment (youknowriad) [reply tools/webpack/packages.js:157]: In other words, will we have to update https://wordpress.org/gutenberg-framework/ as a consequence of this?

Comment (swissspidy) [reply tools/webpack/packages.js:157]: > For me that's a concern, my guess is that we're going to add the vips packages as a dependency of block-editor package.

Indirectly through the new `upload-media` package, yes. Note that first this will be behind an experiment in the plugin only, and wouldn't be exposed through the published npm packages (at least that's my idea). So we have some time to iterate on this. Right now this change doesn't affect any published packages.

> So my question is what happens to third-party block editors, will they have to change how their applications builds, feels heavy? How do regular npm packages (in the npm registry) that use wasm dependencies work.
> In other words, will we have to update https://wordpress.org/gutenberg-framework/ as a consequence of this?

My understanding is that third-party block editor applications will need to update their bundler config to ensure that it supports wasm files. Once this package is published as a script module only, usage might become a bit easier.

Vite (as used by https://wordpress.org/gutenberg-framework/) has some wasm support built-in, but for ESM integration there is `vite-plugin-wasm`

Comment (youknowriad) [reply tools/webpack/packages.js:157]: I wonder if `asset/inline` is actually better for us, that way we can ship npm packages that are ready to use regardless of the bundler config. I'm not sure how I feel about pushing this burden towards consumers. 

Ok for me to start as a Gutenberg experiment not exposed to WordPress and npm users.

Comment (swissspidy) [reply tools/webpack/packages.js:157]: > I wonder if `asset/inline` is actually better for us, that way we can ship npm packages that are ready to use regardless of the bundler config.

That sounds really bad for performance if we shove several MB worth of WASM files as data URIs into the JS bundles... I'd much rather have them as separate files so that they can be cached by the browser or even offloaded to a CDN.

Comment (youknowriad) [reply tools/webpack/packages.js:157]: > That sounds really bad for performance if we shove several MB worth of WASM files as data URIs into the JS bundles... I'd much rather have them as separate files so that they can be cached by the browser or even offloaded to a CDN.

Isn't it the same thing for the JS, it can also be cached and offloaded if needed, not sure I follow what's different here.

**Pull request merged**
